### DISCUSSION TITLE
Can Susan iterate through an array?

### DISCUSSION THREAD

**Discussion by anotheruserofgithub** [OpenModelica/OpenModelica#12596] [Q&A] [answered]
The `|>` operator allows to loop over a `list<>`, but apparently not over an `array<>`. I've tried to search into documentation and user guides, it looks like it has never been fully implemented (for `Array`, at least).

It seems very simple to me, _naïvely_, to generate a MetaModelica traversal function to loop over the elements of an array using `arrayGet()`, so I'm a bit surprised I can't find any example of it.

I'm asking because [`VisualXMLTpl.tpl`](https://github.com/OpenModelica/OpenModelica/blob/d85e7290b4f2a49afa640256db15473e6eda1a49/OMCompiler/Compiler/Template/VisualXMLTpl.tpl) currently has to call `arrayList()` many times for very small arrays (3 or 3x3 elements), where an array traversal operator would do a much more efficient job.

**Comment by anotheruserofgithub** [comment]
Does anyone have an answer? :)

**Comment by perost** [comment]
I don't think Susan supports that. Arrays weren't really used much in the "old days" when Susan was implemented, and there hasn't been much interest in continuing the development since the original developer left unfortunately.

**Reply by anotheruserofgithub** [reply]
Thanks!

**Reply by anotheruserofgithub** [reply, marked as answer]
In the end I decided to implement it by myself in https://github.com/OpenModelica/OpenModelica/pull/12648.

**Comment by anotheruserofgithub** [comment]
Hi @perost, I'm transferring my OM fork onto another machine (Ubuntu 24.04) and trying to compile it with the new `cmake` procedure. The old `make` procedure is still working fine, and if I run `cmake` after `make` it also succeeds (because it reuses some files already compiled by `make`) but if I clean up everything and start with `cmake` it fails with an error:
```
$ cmake -S . -B build_cmake -Wno-dev
$ cmake --build build_cmake --target install --parallel
 
[/home/user/OpenModelica/OMCompiler/Compiler/Template/VisualXMLTpl.tpl:115:40-115:47:writable] Error:
Elaborated expression 'i_element' of type 'array<DAE.Exp>' failed to type adapt to its inferred type 'DAE.Exp'.
```
You won't have this issue on master branch, it's a work in progress on `VisualXML`.

It's obviously the array map operator that is not interpreted correctly. This is confirmed by comparing the generated file `OpenModelica/OMCompiler/Compiler/Template/VisualXMLTpl.mo.err.mo` with what I reported at the end of https://github.com/OpenModelica/OpenModelica/pull/12648#issuecomment-2194427781; with `cmake`:
```modelica
protected function smf_9
  input Tpl.Text in_txt;
  input array<DAE.Exp> in_it;

  output Tpl.Text out_txt;
algorithm
  out_txt :=
  match(in_txt, in_it)
    local
      Tpl.Text txt;
      array<DAE.Exp> i_element;

    case ( txt,
           i_element )
      equation
        txt = dumpExp(txt, #Error# Elaborated expression 'i_element' of type 'array<DAE.Exp>' failed to type adapt to its inferred type 'DAE.Exp'. #(i_element));
      then txt;
  end match;
end smf_9;

public function dumpVecExp
  input Tpl.Text txt;
  input array<DAE.Exp> a_vector;

  output Tpl.Text out_txt;
algorithm
  out_txt := smf_9(txt, a_vector);
end dumpVecExp;
```

I was wondering if this could be due to bootstrapped sources that have not been updated with the changes made by https://github.com/OpenModelica/OpenModelica/pull/12648, or if I would need to change a `cmake` configuration, a bit like `--without-omc` for `make`. Do you know more how both differ in this regard?

I can share the contents of some generated CMake files if you like, but there are so many that I don't know where to check for the answer...

**Reply by perost** [reply]
The TplAbsyn bootstrapping sources where last updated five years ago, so as you say that's most likely the issue. I've started updating them with #14209, if that works fine I'll also update the files in the OMBootstrapping repo used by the cmake build.

We seem to be having some hardware issues with one of our machines at the moment that's causing some issues with our PRs though, so it might take a bit of time before everything is merged.

**Reply by perost** [reply]
Merging went faster than I expected, so the bootstrapping sources should be updated now.

Be aware that you may need to delete the existing sources to trigger a download of the new ones. I think it should be enough to just delete the `OMCompiler/Compiler/boot/bomc` folder.

**Reply by anotheruserofgithub** [reply]
Thanks! I made a clean build and it succeeded. ;)
